readers.py

```diff
@@ -208,48 +218,67 @@ def prepare_reader(self,
+    context_features = {
+        "id": tf.FixedLenFeature([], tf.string),
+    }
+    if self.segment_labels:
+      context_features.update({
+          "segment_labels": tf.VarLenFeature(tf.int64),
+          "segment_start_times": tf.VarLenFeature(tf.int64),
+          "segment_end_times": tf.VarLenFeature(tf.int64),
+          "segment_scores": tf.VarLenFeature(tf.float32)
+      })
+    else:
+      context_features.update({"labels": tf.VarLenFeature(tf.int64)})
+    sequence_features = {
+        feature_name: tf.FixedLenSequenceFeature([], dtype=tf.string)
+        for feature_name in self.feature_names
+    }
     contexts, features = tf.parse_single_sequence_example(
         serialized_example,
-        context_features={"id": tf.FixedLenFeature(
-            [], tf.string),
-                          "labels": tf.VarLenFeature(tf.int64)},
-        sequence_features={
-            feature_name : tf.FixedLenSequenceFeature([], dtype=tf.string)
-            for feature_name in self.feature_names
-        })
-
-    # read ground truth labels
-    labels = (tf.cast(
-        tf.sparse_to_dense(contexts["labels"].values, (self.num_classes,), 1,
-            validate_indices=False),
-        tf.bool))
+        context_features=context_features,
+        sequence_features=sequence_features)
+
+    # Densify frame/segment-level labels.
+    if self.segment_labels:
+      pass
+    else:
+      labels = (
+          tf.cast(
+              tf.sparse_to_dense(
+                  contexts["labels"].values, (self.num_classes,),
+                  1,
+                  validate_indices=False), tf.bool))
 
     # loads (potentially) different types of features and concatenates them
     num_features = len(self.feature_names)
     assert num_features > 0, "No feature selected: feature_names is empty!"
 
     assert len(self.feature_names) == len(self.feature_sizes), \
-    "length of feature_names (={}) != length of feature_sizes (={})".format( \
-    len(self.feature_names), len(self.feature_sizes))
+    "length of feature_names (={}) != length of feature_sizes (={})".format(
+        len(self.feature_names), len(self.feature_sizes))
 
     num_frames = -1  # the number of frames in the video
     feature_matrices = [None] * num_features  # an array of different features
     for feature_index in range(num_features):
       feature_matrix, num_frames_in_this_feature = self.get_video_matrix(
           features[self.feature_names[feature_index]],
-          self.feature_sizes[feature_index],
-          self.max_frames,
-          max_quantized_value,
-          min_quantized_value)
+          self.feature_sizes[feature_index], self.max_frames,
+          max_quantized_value, min_quantized_value)
       if num_frames == -1:
         num_frames = num_frames_in_this_feature
       else:
-        tf.assert_equal(num_frames, num_frames_in_this_feature)
+        tf.assert_equal(num_frames, num_frames_in_this_feature).mark_used()
```

Comment:
I think this is never used because it isn't a required node in the computational graph.

Reply:
Removed.